src/mme/mme-gn-handler.c

```diff
@@ -68,6 +68,117 @@ static int decode_global_enb_id(S1AP_Global_ENB_ID_t *glob_enb_id, const uint8_t
+        return;
+    }
+
+    if (req->ms_validated.presence &&
+        (req->ms_validated.u8 & 0x01) /* 1=> "Yes" */
+        && !req->imsi.presence) {
+        ogs_warn("[Gn] Rx SGSN Context Request with 'MS Validated' but no IMSI!");
+        mme_gtp1_send_sgsn_context_response(NULL, OGS_GTP1_CAUSE_MANDATORY_IE_MISSING, xact);
+        return;
+    }
+
+    rai = req->routeing_area_identity.data;
+
+    if (req->imsi.presence) {
+        mme_ue = mme_ue_find_by_imsi(req->imsi.data, req->imsi.len);
+    } else if (req->packet_tmsi.presence) { /* P-TMSI */
+        if (!req->p_tmsi_signature.presence) {
+            ogs_warn("[Gn] Rx SGSN Context Request with 'P-TMSI' but no P-TMSI Signature! Assuming value 0.");
+            req->p_tmsi_signature.u24 = 0;
+        }
+        rai_ptmsi_to_guti(rai, req->packet_tmsi.u32, req->p_tmsi_signature.u24, &nas_guti);
+        mme_ue = mme_ue_find_by_guti(&nas_guti);
+    } else if (req->temporary_logical_link_identifier.presence) {
+        if (!req->p_tmsi_signature.presence) {
+            ogs_warn("[Gn] Rx SGSN Context Request with 'TLLI' but no P-TMSI Signature! Assuming value 0.");
+            req->p_tmsi_signature.u24 = 0;
+        }
+        /* TS 29.060 7.5.3 "The TLLI/P-TMSI and RAI is a foreign TLLI/P-TMSI and an RAI in the old SGSN."
+         * A foregin TLLI is "tlli = (p_tmsi & 0x3fffffff) | 0x80000000", and since we only use 0x3fffffff
+         * bits of P-TMSI to derive the GUTI, it's totally fine passing the TLLI as P-TMSI. */
+        rai_ptmsi_to_guti(rai, req->temporary_logical_link_identifier.u32, req->p_tmsi_signature.u24, &nas_guti);
+        mme_ue = mme_ue_find_by_guti(&nas_guti);
+    }
+
+    if (!mme_ue) {
+        mme_gtp1_send_sgsn_context_response(NULL, OGS_GTP1_CAUSE_IMSI_IMEI_NOT_KNOWN, xact);
+        return;
+    }
+
+    mme_ue->gn.sgsn_gn_teid = req->tunnel_endpoint_identifier_control_plane.u32;
+
+    rv = ogs_gtp1_gsn_addr_to_ip(req->sgsn_address_for_control_plane.data,
+                                 req->sgsn_address_for_control_plane.len,
+                                 &mme_ue->gn.sgsn_gn_ip);
+    ogs_assert(rv == OGS_OK);
+
+   if (req->alternative_sgsn_address_for_control_plane.presence) {
+        rv = ogs_gtp1_gsn_addr_to_ip(req->alternative_sgsn_address_for_control_plane.data,
+                                     req->alternative_sgsn_address_for_control_plane.len,
+                                    &mme_ue->gn.sgsn_gn_ip_alt);
+        ogs_assert(rv == OGS_OK);
+   }
+
+   mme_gtp1_send_sgsn_context_response(mme_ue, OGS_GTP1_CAUSE_REQUEST_ACCEPTED, xact);
+}
+
+/* TS 29.060 7.5.5 SGSN Context Acknowledge */
+void mme_gn_handle_sgsn_context_acknowledge(
+        ogs_gtp_xact_t *xact, ogs_gtp1_sgsn_context_acknowledge_t *req)
+{
+    /* FIXME: find out what to do here, 3GPP TS 23.401 D.3.5
+     * Probably release the Session on the SGW (without releasing in the PGW) */
```

Comment:
@acetcom I plan to implement this part in a follow-up commit/PR, so that whoever implements SGSN side can already start interacting with open5gs once this PR is merged. in here I will need to implement the following below, any help/advice on how to accomplish it would be welcome. I don't know if there exists already some sort of FSM to tear down the entire bearer/session in both SGW and ENB:

> "The old MME deletes the EPS bearer resources by sending Delete Session Request (Cause; Operation Indication,
> Secondary RAT usage data) messages to the Serving GW". Set the "operation Indication" flag, which "indicates
> to the old Serving GW that the old Serving GW shall not initiate a delete procedure towards the PDN GW."
> Release session at the ENB: "If the old MME has an S1-MME association for the UE, the source MME sends a S1-U Release Command to the
> source eNodeB when receiving the SGSN Context Acknowledge message from the new SGSN. The RRC
> connection is released by the source eNodeB. The source eNodeB confirms the release of the RRC connection
> and of the S1-U connection by sending a S1-U Release Complete message to the source MME."

FYI this comes from 3GPP TS 23.401 Annex D.3.5 "Routing Area Update"